### ISSUE TITLE
Unhandled Behaviour in Loss Calculation causes Error

### ISSUE THREAD

**Issue by cor3ntino** [meituan/YOLOv6#177]
Hello!

I think I noticed an error in the code of `loss.py`. Indeed, the function `dynamic_k_matching` does not handle correclty the case where the model does not predict any box; and such a case happens for some datasets.

I do believe it is the same problem than in the YOLOX library [https://github.com/Megvii-BaseDetection/YOLOX/issues/778](url), and that it could be solved by adding an `if` as below:

```
if n_candidate_k != 0:
            topk_ious, _ = torch.topk(ious_in_boxes_matrix, n_candidate_k, dim=1)
            dynamic_ks = torch.clamp(topk_ious.sum(1).int(), min=1)
            dynamic_ks = dynamic_ks.tolist()

            for gt_idx in range(num_gt):
                _, pos_idx = torch.topk(
                    cost[gt_idx], k=dynamic_ks[gt_idx], largest=False
                )
                matching_matrix[gt_idx][pos_idx] = 1
            del topk_ious, dynamic_ks, pos_idx
```

However, such a fix causes a dimensional issue. Indeed, because of that, `bbox_preds.view(-1, 4)[fg_masks]` and `reg_targets` no longer share the same dimension and thus the IoU loss calculation is impossible line 190 of `loss.py:`
`loss_iou += (self.iou_loss(bbox_preds.view(-1, 4)[fg_masks].T, reg_targets)).sum() / num_fg`

I am guessing that the calculations in the following lines are not possible either for dimensional reasons too.

Would anybody have an idea of how to fix that?

Thanks for your help!

**Comment by cor3ntino** [comment]
Alright, I finally get a way to fix it and created a pull request.

[https://github.com/meituan/YOLOv6/pull/193](url)

This patch fixes the issue of handling cases where model does not predict any box.
This patch does not change the behavior for datasets where the model always makes prediction for every input.

I close the topic hoping it will be merge quickly; be free to re-open it if you think something else is missing.
